File: docs/test-cases/path-traversal.java
// Path traversal — intentionally vulnerable. DO NOT deploy.
import java.nio.file.*;
import org.springframework.web.bind.annotation.*;
import org.springframework.core.io.*;

@RestController
public class FileController {

    private static final String UPLOAD_DIR = "/app/uploads";

    // BUG: Paths.get does not enforce containment
    @GetMapping("/download")
    public Resource download(@RequestParam String filename) {
        Path path = Paths.get(UPLOAD_DIR, filename); // ../../etc/passwd works
        return new FileSystemResource(path.toFile());
    }

    // BUG: no normalization or containment check
    @GetMapping("/read")
    public String readFile(@RequestParam String path) throws Exception {
        Path full = Paths.get(UPLOAD_DIR).resolve(path);
        // resolve doesn't prevent traversal — attacker sends ../../../etc/shadow
        return Files.readString(full);
    }

    // BUG: user controls subdirectory and filename
    @DeleteMapping("/files/{dir}/{name}")
    public String deleteFile(@PathVariable String dir, @PathVariable String name)
            throws Exception {
        Path target = Paths.get(UPLOAD_DIR, dir, name);
        Files.delete(target); // attacker deletes ../../config/application.yml
        return "deleted";
    }
}

File: docs/test-cases/path-traversal.py
"""Path traversal — intentionally vulnerable. DO NOT deploy."""
import os
from flask import Flask, request, send_file

app = Flask(__name__)
UPLOAD_DIR = "/app/uploads"

# BUG: user-supplied filename allows ../../../etc/passwd
@app.route("/download")
def download():
    filename = request.args.get("file")
    path = os.path.join(UPLOAD_DIR, filename)  # join doesn't prevent traversal
    return send_file(path)

# BUG: absolute path ignores base directory entirely
@app.route("/read")
def read_file():
    filepath = request.args.get("path")
    full = os.path.join(UPLOAD_DIR, filepath)  # /etc/passwd ignores UPLOAD_DIR
    with open(full) as f:
        return f.read()

# BUG: symlink escape — attacker uploads symlink pointing outside root
@app.route("/serve/<name>")
def serve(name):
    path = os.path.join(UPLOAD_DIR, name)
    # No symlink check — symlink could point to /etc/shadow
    return send_file(path)
